// @googlemaps/google-maps-services-js/dist/serialize.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const query_string_1 = require("query-string");
const util_1 = require("./util");
const separator = "|";
function latLngToString(o) {
    if (typeof o === "string") {
        return o;
    }
    else if (Array.isArray(o) && o.length === 2) {
        // no transformation
    }
    else if ("lat" in o && "lng" in o) {
        o = [o.lat, o.lng];
    }
    else if ("latitude" in o && "longitude" in o) {
        o = [o.latitude, o.longitude];
    }
    else {
        throw new TypeError();
    }
    return o
        .map(x => {
        return x.toString();
    })
        .join(",");
}
exports.latLngToString = latLngToString;
function objectToString(o) {
    if (typeof o === "string") {
        return o;
    }
    else {
        let keys = Object.keys(o);
        keys.sort();
        return keys.map(k => k + ":" + o[k]).join(separator);
    }
}
exports.objectToString = objectToString;
function latLngBoundsToString(latLngBounds) {
    if (typeof latLngBounds === "string") {
        return latLngBounds;
    }
    else {
        return (latLngToString(latLngBounds.southwest) +
            separator +
            latLngToString(latLngBounds.northeast));
    }
}
exports.latLngBoundsToString = latLngBoundsToString;
function toLatLngLiteral(o) {
    if (typeof o === "string") {
        const parts = o.split(",").map(Number);
        return { lat: parts[0], lng: parts[1] };
    }
    else if (Array.isArray(o) && o.length === 2) {
        const parts = o.map(Number);
        return { lat: parts[0], lng: parts[1] };
    }
    else if ("lat" in o && "lng" in o) {
        return o;
    }
    else if ("latitude" in o && "longitude" in o) {
        return { lat: o.latitude, lng: o.longitude };
    }
    else {
        throw new TypeError();
    }
}
exports.toLatLngLiteral = toLatLngLiteral;
function latLngArrayToStringMaybeEncoded(o) {
    if (typeof o === "string") {
        return o;
    }
    const concatenated = o.map(latLngToString).join(separator);
    const encoded = `enc:${util_1.encodePath(o.map(toLatLngLiteral))}`;
    if (encoded.length < concatenated.length) {
        return encoded;
    }
    return concatenated;
}
exports.latLngArrayToStringMaybeEncoded = latLngArrayToStringMaybeEncoded;
function serializer(format, queryStringOptions = {
    arrayFormat: "separator",
    arrayFormatSeparator: separator
}) {
    return (params) => {
        Object.keys(format).forEach((key) => {
            if (key in params) {
                params[key] = format[key](params[key]);
            }
        });
        return query_string_1.stringify(params, queryStringOptions);
    };
}
exports.serializer = serializer;
